### PR TITLE
fix(DQL): optimize query for has function with offset. (#7727)

### PR DESCRIPTION
**Pull request by all-seeing-code**
Fixes DGRAPH-574.

This PR reduces the latency of the `has` function slightly. For example the given query:-
```
{
    me(func: has(actor.film), first: 10, offset: 300000){
       name@en
    }
}
```
when run on the `21million` dataset takes around `80` milliseconds to execute as compared to earlier around `93` milliseconds.

Machine Specs:-
```
Model Name:	MacBook Pro
  Model Identifier:	MacBookPro16,1
  Processor Name:	6-Core Intel Core i7
  Processor Speed:	2.6 GHz
  Number of Processors:	1
  Total Number of Cores:	6
  L2 Cache (per Core):	256 KB
  L3 Cache:	12 MB
  Memory:	16 GB
```

Note:-
This does not optimize for the `has` function on predicates with the `@lang` tag.

(cherry picked from commit 9ba15b79da1900344ff95d87b127b6676853d257)

### CODE REVIEW

**Review by MichelDiz — APPROVED** [review]

**Review by matthewmcneely — APPROVED** [review]

**Comment by MichelDiz** [comment]
@anurags92 this needs to clean up the hidden comments.

File: query/query.go [hunk 1]
```diff
@@ -2256,10 +2258,14 @@ func ProcessGraph(ctx context.Context, sg, parent *SubGraph, rch chan error) {
 	}
 
 	if len(sg.Params.Order) == 0 && len(sg.Params.FacetsOrder) == 0 {
-		// There is no ordering. Just apply pagination and return.
-		if err = sg.applyPagination(ctx); err != nil {
-			rch <- err
-			return
+		// for `has` function when there is no filtering and ordering, we fetch
+		// correct paginated results so no need to apply pagination here.
+		if !(len(sg.Filters) == 0 && sg.SrcFunc != nil && sg.SrcFunc.Name == "has") {
```
Comment (mangalaman93) [inline query/query.go:2263]: Is there a case possible where this if evaluates to false but we still need to perform pagination like before?

File: query/query.go [hunk 2]
```diff
@@ -991,12 +993,12 @@ func calculateFirstN(sg *SubGraph) int32 {
...
 	if len(sg.Filters) == 0 && len(sg.Params.Order) == 0 && !shouldExclude {
-		// Offset also added because, we need n results to trim the offset.
 		if sg.Params.Count != 0 {
-			count = sg.Params.Count + sg.Params.Offset
+			return int32(sg.Params.Count), int32(sg.Params.Offset)
 		}
 	}
-	return int32(count)
+	// make offset = 0, if there is need to fetch all the results.
+	return int32(count), 0
```
Comment (akon-dey) [inline query/query.go:1001]: This should probably be:
        return int32(count), int32(sg.Params.Offset)
There is a case where one might want all the results after skipping the first few (denoted by the Offset).

Comment (matthewmcneely) [reply query/query.go:1001]: @akon-dey Actually no, I think that in this case worker/task.go needs `pb.query.Offset` to in fact be 0 in order to function correctly.

**Comment by anurags92** [comment]
This PR has a failing test, which needs to be properly understood. I am marking this as draft for now.

**Comment by anurags92** [comment]
I don't see a significant performance improvement on my end.

`main`
```
Dgraph version   : local
Dgraph codename  : dgraph
Dgraph SHA-256   : 8525cf5541e0d9d63398304254cc9638d92a94b4e0dece17a1b0212d509aac81
Commit SHA-1     : a4729b3ca
Commit timestamp : 2022-11-30 17:57:07 -0300
Branch           : main
Go version       : go1.19.2
jemalloc enabled : true
```
Benchmark results:
```
goos: linux
goarch: amd64
pkg: github.com/dgraph-io/test
cpu: 12th Gen Intel(R) Core(TM) i7-12700H
BenchmarkQuery100_0-20               403           2635436 ns/op
BenchmarkQuery100_20-20              444           3041838 ns/op
BenchmarkQuery100_200-20             336           3470780 ns/op
BenchmarkQuery100_20000-20           134           7816592 ns/op
PASS
ok      github.com/dgraph-io/test       7.107s
```

`cherry-pick-7727`
```
Dgraph version   : local
Dgraph codename  : dgraph
Dgraph SHA-256   : 23a5ae17dfef9391a7131d1953595e30f5a919f5db61ba9635ac6a1bd7383432
Commit SHA-1     : 5e6f44c03
Commit timestamp : 2022-12-01 17:31:05 +0530
Branch           : cherry-pick-7727
Go version       : go1.19.2
jemalloc enabled : true
```
Benchmark results:
```
goos: linux
goarch: amd64
pkg: github.com/dgraph-io/test
cpu: 12th Gen Intel(R) Core(TM) i7-12700H
BenchmarkQuery100_0-20               391           3201947 ns/op
BenchmarkQuery100_20-20              363           3034455 ns/op
BenchmarkQuery100_200-20             481           2211461 ns/op
BenchmarkQuery100_20000-20           229           8433493 ns/op
PASS
ok      github.com/dgraph-io/test       6.790s
```


Here's [the code](https://go.dev/play/p/mPp_gnl6Gcg) I used to get the results. It expects a local dgraph running with `ldbc-sf.03` dataset inserted.

**Comment by anurags92** [comment]
> This PR has a failing test, which needs to be properly understood. I am marking this as draft for now.

This was fixed. It had a test which required a different set of changes. They are being brought in via: https://github.com/dgraph-io/dgraph/pull/8471

**Comment by MichelDiz** [comment]
@anurags92 the description  have some hidden comments. Please, remove it.

**Comment by anurags92** [comment]
I see considerable perfomance improvement on a larger dataset (sf10).

`main`
```
alvis@alvis-alien-ubuntu:~/go/src/github.com/dgraph-io/test$ go test -run=XXX -bench=BenchmarkQuery
goos: linux
goarch: amd64
pkg: github.com/dgraph-io/test
cpu: 12th Gen Intel(R) Core(TM) i7-12700H
BenchmarkQuery100_0-20               502           2370298 ns/op
BenchmarkQuery100_20-20              696           2391215 ns/op
BenchmarkQuery100_200-20             552           2456952 ns/op
BenchmarkQuery100_20000-20           187           7336503 ns/op
PASS
ok      github.com/dgraph-io/test       8.886s
```
`cherry-pick-7727`
```
alvis@alvis-alien-ubuntu:~/go/src/github.com/dgraph-io/test$ go test -run=XXX -bench=BenchmarkQuery
goos: linux
goarch: amd64
pkg: github.com/dgraph-io/test
cpu: 12th Gen Intel(R) Core(TM) i7-12700H
BenchmarkQuery100_0-20               470           2335904 ns/op
BenchmarkQuery100_20-20              489           2371475 ns/op
BenchmarkQuery100_200-20             469           2380480 ns/op
BenchmarkQuery100_20000-20           166           6956870 ns/op
PASS
ok      github.com/dgraph-io/test       5.815s
```

**Pull request merged**